nixos/borgbackup: Add option createCommand

dotdoom opened the pull request:
Add an option `service.borgbackup.jobs.<name>.createCommand` that allows using `borg import-tar` rather than `borg create` to import an archive.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dotdoom committed:
nixos/borgbackup: Add option dumpCommandProducesTar

Add an option `service.borgbackup.jobs.<name>.dumpCommandProducesTar`
that allows using `borg import-tar` rather than `borg create` to import
an archive.

dotdoom commented:
Hello @dotlambda and @Scrumplex, I'm looking for maintainers to take a look at this humble PR, or trigger the checks workflow!

Scrumplex reviewed (commented):
Changes seem good. I think the naming of the option is a little confusing if you are not aware of what import-tar does :O

dotdoom commented:
Agreed.  My mind is already spoiled by years of using "borg import-tar".  If you happen to have a suggestion, it would be very welcome!

Scrumplex commented:
I think an assertion would be best here. If these option values are not used/supported/implemented for any reason, then users should make sure they are unset to avoid confusion.

dotdoom committed:
nixos/borgbackup: Add option createCommand

Add an option `service.borgbackup.jobs.<name>.createCommand`
that allows using `borg import-tar` rather than `borg create` to import
an archive.

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -310,6 +315,23 @@ let
   };
 
+  mkCreateCommandImportTarDumpCommandAssertion = name: cfg: {
+    assertion = cfg.createCommand != "import-tar" || cfg.dumpCommand != null;
```
```suggestion
    assertion = cfg.createCommand == "import-tar" -> cfg.dumpCommand != null;
```
is logically equivalent but much more readable

dotdoom commented:
Thank you @Scrumplex for swift review and actionable feedback!

Thanks @dotlambda for suggesting implication operator, it does indeed looks cleaner that way -- I will follow up with a PR to update these.
